### PR TITLE
TF serving support request logging

### PR DESCRIPTION
**Pull request by lluunn**
For #1000 

/cc @jlewi

- Using fluentd as a sidecar container to export logs.
- Choosing bigquery to begin with. Should be easy to add other backends using fluentd plugins.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/1229)

### CODE REVIEW

**Comment by lluunn** [comment]
/hold
need md

**Comment by jlewi** [comment]
Please update the PR description to describe the approach being taken in this PR to solve the request logging problem.

File: components/k8s-model-server/request-logger/logging_worker.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  table_ref = dataset_ref.table(args.table)
+  table = client.get_table(table_ref)
+
+  # rotating files.
+  target_file = args.file_path + '.1'
+  while True:
+    time.sleep(10)
+    if not os.path.isfile(target_file):
+      continue
+    print('found the file\n')
```
Comment (jlewi) [inline components/k8s-model-server/request-logger/logging_worker.py]: Use logging not print.

File: components/k8s-model-server/request-logger/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+google-cloud-bigquery
```
Comment (jlewi) [inline components/k8s-model-server/request-logger/requirements.txt]: Why do you need bigquery?

File: components/k8s-model-server/request-logger/logging_worker.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+"""
+from __future__ import print_function
+
+import argparse
+import ast
+import json
+import os
+import time
+
+from google.cloud import bigquery
```
Comment (jlewi) [inline components/k8s-model-server/request-logger/logging_worker.py]: Why do we need to write our own logging agent as opposed to using fluentd?
For example, https://github.com/kaizenplatform/fluent-plugin-bigquery

File: components/k8s-model-server/http-proxy/server.py [hunk 4]
```diff
@@ -19,10 +19,12 @@
 from itertools import repeat
 import base64
 import logging
-import grpc
+import random
```
Comment (jlewi) [inline components/k8s-model-server/http-proxy/server.py]: If we do it in the http server proxy we won't get this for gRPC requests; right?

**Comment by jlewi** [comment]
Are you emitting the logs in json format?

**Comment by lluunn** [comment]
Using fluentd now, PTAL
@jlewi

**Comment by lluunn** [comment]
/hold cancel

It's working now

**Comment by jlewi** [comment]
This is great.

Can you update the PR description to explain the design choices; i.e. the fact that we are using fluentd and BigQuery?

**Comment by lluunn** [comment]
Some of the comment are regarding how we present the prototype:
1. use tf-serving package, setting parameters, adding volumes/env if XXX
2. use examples package, ks generate copies the whole jsonnet, and let users modify directly.

I think the current libsonnet is already messy (filed #1264 ), so I choose option 2 here.

WDYT?

**Comment by jlewi** [comment]
>use tf-serving package, setting parameters, adding volumes/env if XXX
use examples package, ks generate copies the whole jsonnet, and let users modify directly.

The prototype belongs in the tf-serving package regardless of how you generate it; i.e. you can still copy the whole jsonnet.

Telling a user to change certain lines in a text file is much less convenient then doing

```
ks param set fluent-conf project myproject
```

Since these parameters are required we should try to expose them as `ks param set` commands rather than requiring users to modify the conf.

**Comment by lluunn** [comment]
Remove the conf file and make it in the prototype. project etc can be set via `ks param set`.
PTAL, thanks!

**Comment by lluunn** [comment]
test failure: #1266

File: kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+              },
+              requests: {
+                cpu: "1",
+                memory: "1Gi",
+              },
+            },
+          },
+          // Http proxy
+          {
+            name: "mnist-http-proxy",
```
Comment (jlewi) [inline kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet]: remove mnist in the name.

File: kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+              fsGroup: 1000,
+            },
+            volumeMounts: [
+              {
+                name: "request-logs",
+                mountPath: "/tmp/logs",
+              },
+            ],
+          },
+          // Logging container.
```
Comment (jlewi) [inline kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet]: Lets file an issue to create an admission controller to inject the side car.

File: kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+          {
+            name: "mnist-logging",
```
Comment (jlewi) [inline kubeflow/tf-serving/prototypes/tf-serving-with-request-log.jsonnet]: Don't use mnist in the name.

**Comment by jlewi** [comment]
Sorry for the slow reply; GCP next.

Minor comment about replacing the name "mnist" somewhere.

The duplication of code with the other templates is unfortunate. I think the best option is probably to create an admission controller that can inject the relevant information and then just make that its own prototype. Lets file a follow on issue.

**Pull request merged**